### PR TITLE
HTTP/2 goaway connection state update sequencing

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
The Http2Connection state is updated by the DefaultHttp2ConnectionDecoder after the frame listener is notified of the goaway frame. If the listener sends a frame synchronously this means the connection state will not know about the goaway it just received and we may send frames that are not allowed on the connection. This may also mean a stream object is created but it may never get taken out of the stream map unless some other event occurs (e.g. timeout).

Modifications:
- The Http2Connection state should be updated before the listener is notified of the goaway
- The Http2Connection state modification and validation should be self contained when processing a goaway instead of partially in the decoder.

Result:
No more creating streams and sending frames after a goaway has been sent or received.

### CODE REVIEW

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Connection.java [hunk 1]
```diff
@@ -327,15 +327,17 @@
...
     boolean goAwaySent();
 
     /**
-     * Indicates that a {@code GOAWAY} was sent to the remote endpoint and sets the last known stream.
+     * Updates the local state of this {@link Http2Connection} as a result of a {@code GOAWAY} to send to the remote
+     * endpoint.
+     * @return {@code true} if the corresponding {@code GOAWAY} frame should be sent to the remote endpoint.
      */
-    void goAwaySent(int lastKnownStream, long errorCode, ByteBuf message);
+    boolean goAwaySent(int lastKnownStream, long errorCode, ByteBuf message) throws Http2Exception;
```
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Connection.java]: note this API change. I think it makes more sense to have the connection validate these conditions, and is consistent with other connection level validation. Having the validation outside can lead to the issue fixed in this PR (the state was updated at the wrong time).

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java [hunk 2]
```diff
@@ -225,7 +225,11 @@ public boolean goAwayReceived() {
     }
 
     @Override
-    public void goAwayReceived(final int lastKnownStream, long errorCode, ByteBuf debugData) {
+    public void goAwayReceived(final int lastKnownStream, long errorCode, ByteBuf debugData) throws Http2Exception {
+        if (localEndpoint.lastStreamKnownByPeer() >= 0 && localEndpoint.lastStreamKnownByPeer() < lastKnownStream) {
+            throw connectionError(PROTOCOL_ERROR, "lastStreamId MUST NOT increase. Current value: %d new value: %d",
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: I am correct that the debugData does not need to be released here as the ownership of the buffer is not transferred ?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java [hunk 3]
```diff
@@ -256,7 +256,20 @@ public boolean goAwaySent() {
...
-    public void goAwaySent(final int lastKnownStream, long errorCode, ByteBuf debugData) {
+    public boolean goAwaySent(final int lastKnownStream, long errorCode, ByteBuf debugData) throws Http2Exception {
+        if (remoteEndpoint.lastStreamKnownByPeer() >= 0) {
+            // Protect against re-entrancy. Could happen if writing the frame fails, and error handling
+            // treating this is a connection handler and doing a graceful shutdown...
+            if (lastKnownStream == remoteEndpoint.lastStreamKnownByPeer()) {
+                return false;
+            }
+            if (lastKnownStream > remoteEndpoint.lastStreamKnownByPeer()) {
+                throw connectionError(PROTOCOL_ERROR, "Last stream identifier must not increase between " +
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: I am correct that the debugData does not need to be released here as the ownership of the buffer is not transferred ?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java [hunk 4]
```diff
@@ -266,19 +279,16 @@ public void goAwaySent(final int lastKnownStream, long errorCode, ByteBuf debugD
...
-            forEachActiveStream(new Http2StreamVisitor() {
-                @Override
-                public boolean visit(Http2Stream stream) {
-                    if (stream.id() > lastKnownStream && remoteEndpoint.isValidStreamId(stream.id())) {
-                        stream.close();
-                    }
-                    return true;
+        forEachActiveStream(new Http2StreamVisitor() {
+            @Override
+            public boolean visit(Http2Stream stream) {
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: Seems like this is the same code as https://github.com/netty/netty/pull/8080/files#diff-9d991ca64301bcd5b3c00d053591f485R244 . Can we just create an `HttpStreamVisitor` imp that we share  by passing the needed stuff via the constructor ?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java [hunk 5]
```diff
@@ -541,6 +537,7 @@ private boolean shouldIgnoreHeadersOrDataFrame(ChannelHandlerContext ctx, int st
                                 ("Stream created after GOAWAY sent. Last known stream by peer " +
                                  connection.remote().lastStreamKnownByPeer()));
                 }
+                stream.close();
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java]: can you add a comment why this was added ?

File: codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionTest.java [hunk 6]
```diff
@@ -425,7 +425,7 @@ public void reserveWithPushDisallowedShouldThrow() throws Http2Exception {
     @Test(expected = Http2Exception.class)
     public void goAwayReceivedShouldDisallowCreation() throws Http2Exception {
         server.goAwayReceived(0, 1L, Unpooled.EMPTY_BUFFER);
-        server.remote().createStream(3, true);
+        server.local().createStream(3, true);
```
Comment (normanmaurer) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionTest.java]: why this change ?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 2 above)
Comment (Scottmitch) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: yes. I added a comment to the javadocs of the interface.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 3 above)
Comment (Scottmitch) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: yes. I added a comment to the javadocs of the interface.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java (hunk 4 above)
Comment (Scottmitch) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java]: I extracted the iteration into a shared method.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java (hunk 5 above)
Comment (Scottmitch) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java]: yip done.

File: codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionTest.java (hunk 6 above)
Comment (Scottmitch) [reply codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionTest.java]: the unit test was incorrect and highlighted a bug addressed by this PR.

The test simulates receiving a go away frame, then created a stream from the remote endpoint, and expected an error. However this is not correct. Receiving a goaway frame should prevent the *local* endpoint from creating streams and have no effect on the *remote* endpoint.

Comment (Scottmitch) [reply codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionTest.java]: I added some additional tests to catch all the cases and clarify.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java [hunk 7]
```diff
@@ -541,6 +537,10 @@ private boolean shouldIgnoreHeadersOrDataFrame(ChannelHandlerContext ctx, int st
                                 ("Stream created after GOAWAY sent. Last known stream by peer " +
                                  connection.remote().lastStreamKnownByPeer()));
                 }
+
+                // If we are ignoring frames for this stream it means that we no longer need to retain state for this
+                // stream and can close it, which should remove it from the active stream map.
+                stream.close();
```
Comment (bryce-anderson) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java]: It doesn't seem like the responsibility of the decoder to manage the streams life-cycle Under what conditions would this happen? It seems that sending a RST should always be coupled with closing the stream and you've covered the GOAWAY case above. From my limited digging, it looks like the [`Http2ConnectionHandler`](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java#L778) is the only way we'll hit the `isResetSent()` case and it's already setup a listener to close the session. I presume it's only delayed to make draining work correctly with resets.

Comment (normanmaurer) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java]: @Scottmitch can you chime in here ?

Comment (Scottmitch) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java]: This was added as a fail-safe because prior to this PR we would still create streams initiated by the remote endpoint even if we sent an goaway. This resulted in "leaked" stream objects (stream objects would be created, inserted into the `streamMap` but never removed because there would be no lifetime event to trigger removal). With this PR `DefaultHttp2Connection` no longer permits this and I don't think we should hit this code path because:

1. as you said sending a rst should close the stream, which will remove from the `streamMap`.
2. we should no longer create streams initiated by the remote peer (exceeding the last stream ID) after a go away was sent.

so I can remove this code block if we assume this scenario is invalid for any implementation of `Http2Connection`. We have unit tests which do exercise this code path in `DefaultHttp2ConnectionDecoderTest` but we can modify the test/mocks to not

Comment (bryce-anderson) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java]: I can't attest to other implementations of `Http2Connection`, but if that's a legit concern I don't mind if we leave it if the worst case scenario is it's dead code. I'll trust your judgment @Scottmitch.

Comment (Scottmitch) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java]: I'm not aware of any alternative implementation of `Http2Connection` and you have a good point this code change seems to be fixing an issue that stems from elsewhere. Now that `DefaultHttp2Connection` is fixed it should be sufficient to remove this with Netty's code, but I'll add a comment to clarify the assumptions.
